gateway/src/apicast/policy_loader.lua

```diff
@@ -0,0 +1,199 @@
+local root_loaded = package.loaded
+
+local root_require = require
+local pcall = pcall
+
+local preload = package.preload
+
+--- create a require function not using the global namespace
+-- loading code from policy namespace should have no effect on the global namespace
+-- but poliocy can load shared libraries that would be cached globally
+local function gen_require(package)
+
+  local function not_found(modname, err)
+    return error(format("module '%s' not found:%s", modname, err), 0)
+  end
+
+  --- helper function to safely use the native require function
+  local function fallback(modname)
+    local mod
+
+    mod = package.loaded[modname]
+
+    if not mod then
+      ngx.log(ngx.DEBUG, 'native require for: ', modname)
+      mod = root_require(modname)
+    end
+
+    return mod
+  end
+
+  --- helper function to find and return correct loader for a module
+  local function find_loader(modname)
+    local loader, file, err, ret
+
+    -- http://www.lua.org/manual/5.2/manual.html#pdf-package.searchers
+
+    -- When looking for a module, require calls each of these searchers in ascending order,
+    for i=1, #package.searchers do
+      -- with the module name (the argument given to require) as its sole parameter.
+      ret, err = package.searchers[i](modname)
+
+      -- The function can return another function (the module loader)
+      -- plus an extra value that will be passed to that loader,
+      if type(ret) == 'function' then
+        loader = ret
+        file = err
+        break
+      -- or a string explaining why it did not find that module
+      elseif type(ret) == 'string' then
+        err = ret
+      end
+      -- (or nil if it has nothing to say).
+    end
+
+    return loader, file, err
+  end
+
+  --- reimplemented require function
+  -- - return a module if it was already loaded (globally or locally)
+  -- - try to find loader function
+  -- - fallback to global require
+  -- @tparam string modname module name
+  -- @tparam boolean exclusive load only policy code, turns off the fallback loader
+  return function(modname, exclusive)
+    ngx.log(ngx.DEBUG, 'sandbox require: ', modname)
+    local mod = root_loaded[modname]
+
+    if mod then return mod end
+
+    local loader, file, err = find_loader(modname)
+
+    if loader then
+      ngx.log(ngx.DEBUG, 'sandboxed require for: ', modname, ' file: ', file)
+      mod = loader(modname, file)
+    elseif not exclusive then
+      ngx.log(ngx.DEBUG, 'fallback loader for: ', modname, ' error: ', err)
+      mod = fallback(modname)
+    else
+      return not_found(modname, err)
+    end
+
+    if mod ~= nil then
+      package.loaded[modname] = mod
+    else
+      package.loaded[modname] = true
+    end
+
+    return package.loaded[modname]
+  end
+end
+
+--- this is environment exposed to the policies
+-- that means this is very light sandbox so policies don't mutate global env
+-- and most importantly we replace the require function with our own
+_M.env = {
```

Comment:
Just curious, how did you decide what to include and what not to include here ?

Reply:
I want to include everything that is in standard Lua env. And it was just trial & error. So I guess something might be missing still.
I could go through every key in the normal env and make sure it is here.

Reply:
We probably don't want to expose code loading functions that could escape the sandbox, right?
The sandbox is meant to be just light shell, not a thorough security measure, but still.

Here is global from repl (without package):
```lua
<1>{
  _ = <table 1>,
  _G = <table 1>,
  _VERSION = "Lua 5.1",
  assert = <function 1>,
  bit = {
    arshift = <function 2>,
    band = <function 3>,
    bnot = <function 4>,
    bor = <function 5>,
    bswap = <function 6>,
    bxor = <function 7>,
    lshift = <function 8>,
    rol = <function 9>,
    ror = <function 10>,
    rshift = <function 11>,
    tobit = <function 12>,
    tohex = <function 13>
  },
  collectgarbage = <function 14>,
  coroutine = {
    create = <function 15>,
    isyieldable = <function 16>,
    resume = <function 17>,
    running = <function 18>,
    status = <function 19>,
    wrap = <function 20>,
    yield = <function 21>
  },
  debug = {
    debug = <function 22>,
    getfenv = <function 23>,
    gethook = <function 24>,
    getinfo = <function 25>,
    getlocal = <function 26>,
    getmetatable = <function 27>,
    getregistry = <function 28>,
    getupvalue = <function 29>,
    getuservalue = <function 30>,
    setfenv = <function 31>,
    sethook = <function 32>,
    setlocal = <function 33>,
    setmetatable = <function 34>,
    setupvalue = <function 35>,
    setuservalue = <function 36>,
    traceback = <function 37>,
    upvalueid = <function 38>,
    upvaluejoin = <function 39>
  },
  dofile = <function 40>,
  error = <function 41>,
  gcinfo = <function 42>,
  getfenv = <function 43>,
  getmetatable = <function 44>,
  io = {
    close = <function 45>,
    flush = <function 46>,
    input = <function 47>,
    lines = <function 48>,
    open = <function 49>,
    output = <function 50>,
    popen = <function 51>,
    read = <function 52>,
    stderr = <userdata 1>,
    stdin = <userdata 2>,
    stdout = <userdata 3>,
    tmpfile = <function 53>,
    type = <function 54>,
    write = <function 55>
  },
  ipairs = <function 56>,
  jit = {
    arch = "x64",
    attach = <function 57>,
    flush = <function 58>,
    off = <function 59>,
    on = <function 60>,
    opt = {
      start = <function 61>
    },
    os = "OSX",
    status = <function 62>,
    version = "LuaJIT 2.1.0-beta3",
    version_num = 20100
  },
  load = <function 63>,
  loadfile = <function 64>,
  loadstring = <function 65>,
  math = {
    abs = <function 66>,
    acos = <function 67>,
    asin = <function 68>,
    atan = <function 69>,
    atan2 = <function 70>,
    ceil = <function 71>,
    cos = <function 72>,
    cosh = <function 73>,
    deg = <function 74>,
    exp = <function 75>,
    floor = <function 76>,
    fmod = <function 77>,
    frexp = <function 78>,
    huge = inf,
    ldexp = <function 79>,
    log = <function 80>,
    log10 = <function 81>,
    max = <function 82>,
    min = <function 83>,
    modf = <function 84>,
    pi = 3.1415926535898,
    pow = <function 85>,
    rad = <function 86>,
    random = <function 87>,
    randomseed = <function 88>,
    sin = <function 89>,
    sinh = <function 90>,
    sqrt = <function 91>,
    tan = <function 92>,
    tanh = <function 93>
  },
  module = <function 94>,
  newproxy = <function 95>,
  next = <function 96>,
  os = {
    clock = <function 97>,
    date = <function 98>,
    difftime = <function 99>,
    execute = <function 100>,
    exit = <function 101>,
    getenv = <function 102>,
    remove = <function 103>,
    rename = <function 104>,
    setlocale = <function 105>,
    time = <function 106>,
    tmpname = <function 107>
  },
  pairs = <function 108>,
  pcall = <function 109>,
  print = <function 110>,
  rawequal = <function 111>,
  rawget = <function 112>,
  rawlen = <function 113>,
  rawset = <function 114>,
  require = <function 115>,
  select = <function 116>,
  setfenv = <function 117>,
  setmetatable = <function 118>,
  string = {
    byte = <function 119>,
    char = <function 120>,
    dump = <function 121>,
    find = <function 122>,
    format = <function 123>,
    gmatch = <function 124>,
    gsub = <function 125>,
    len = <function 126>,
    lower = <function 127>,
    match = <function 128>,
    rep = <function 129>,
    reverse = <function 130>,
    sub = <function 131>,
    upper = <function 132>
  },
  table = {
    concat = <function 133>,
    foreach = <function 134>,
    foreachi = <function 135>,
    getn = <function 136>,
    insert = <function 137>,
    maxn = <function 138>,
    move = <function 139>,
    pack = <function 140>,
    remove = <function 141>,
    sort = <function 142>,
    unpack = <function 143>
  },
  tonumber = <function 144>,
  tostring = <function 145>,
  type = <function 146>,
  unpack = <function 143>,
  xpcall = <function 147>
}
```

Reply:
Made a new sandbox in https://github.com/3scale/apicast/pull/566/commits/6884ae0a6c78e024fb890e2fc245b2d8da10f2b9